Sources/App/Scenes/CarPlaySceneDelegate.swift

```diff
@@ -0,0 +1,261 @@
+import CarPlay
+import Communicator
+import HAKit
+import PromiseKit
+import Shared
+
+
+public protocol EntitiesStateSubscription {
+    func subscribe()
+    func unsubscribe()
+}
+
+@available(iOS 16.0, *)
+class CarPlayDelegate: UIResponder {
+    public static let SUPPORTED_DOMAINS_WITH_STRING = [
+        "button": L10n.Carplay.Labels.buttons,
+        "cover": L10n.Carplay.Labels.covers,
+        "input_boolean": L10n.Carplay.Labels.inputBooleans,
+        "input_button": L10n.Carplay.Labels.inputButtons,
+        "light": L10n.Carplay.Labels.lights,
+        "lock": L10n.Carplay.Labels.locks,
+        "scene": L10n.Carplay.Labels.scenes,
+        "script": L10n.Carplay.Labels.scripts,
+        "switch": L10n.Carplay.Labels.switches
+    ]
+    
+    public let SUPPORTED_DOMAINS = SUPPORTED_DOMAINS_WITH_STRING.keys
+    
+    private var MAP_DOMAINS = [
+        "device_tracker",
+        "person",
+        "sensor",
+        "zone"
+    ]
+    
+    private var interfaceController: CPInterfaceController?
+    private var filteredEntities: [HAEntity] = []
+    private var entitiesGridTemplate: EntitiesGridTemplate?
+    private var domainsListTemplate: DomainsListTemplate?
+    private var entitiesStateSubscribeCancelable: HACancellable?
+    private var serverObserver: HACancellable?
+    private var serverId: Identifier<Server>? {
+        didSet {
+            loadEntities()
+        }
+    }
+    
+    let prefs = UserDefaults(suiteName: Constants.AppGroupID)!
+    
+    func loadEntities() {
+        self.domainsListTemplate?.setServerListButton(show: Current.servers.all.count > 1)
+        
+        guard let serverId = serverId, let server = Current.servers.server(for: serverId) else {
+            Current.Log.info("No server available to get entities")
+            filteredEntities.removeAll()
+            self.domainsListTemplate?.entitiesUpdate(updateEntities: filteredEntities)
+            return
+        }
+        
+        guard let allServerEntities = Current.api(for: server).connection.caches.states.value?.all else {
+            Current.Log.info("No entities available from server \(server.info.name)")
+            filteredEntities.removeAll()
+            self.domainsListTemplate?.entitiesUpdate(updateEntities: filteredEntities)
+            interfaceController?.setRootTemplate(self.domainsListTemplate!.getTemplate(), animated: false)
+            return
+        }
+        
+        filteredEntities = getFilteredAndSortEntities(entities: Array(allServerEntities))
```

Comment:
pass around a `HACache` instance, and don't worry about passing updated when it changes. for example, `connection.caches.states.map { /* turn into your filtered version */ }` gives you `HACache<YourTypeValue>` which live-updates when changes occur.

Reply:
I renamed the method and removed the sort that existed. This method is filtering the entities by the types that are currently supported by the CarPlay App. If I add the code you suggested, it will include all types of entities, even those that are not supported. This can create problems in the view that shows the entities from the selected domain.

Is this change ok, or is there a better way to apply this filter?

Reply:
See the snippet above. You should not be passing around an `Array` of entities like this; the `HACache` will give you a live-updating reference to the filtered version if you use the method above.